### PR TITLE
feat(rich-text-editor) - single click actions 

### PR DESCRIPTION
**Pull request by gavin-everett-genesys**
This change includes the following,

1. Addition of actions which include **bold**, **italic**, **underline** and **strikethrough**, **undo**, **redo**, **clear formatting**, **delete**, **bullet list**, **ordered list**, **block quote**
2. Also includes the **disabled** state and the optional action divider.

Next steps :

1. I am going to update this PR to include all the actions that just require a single click since its pretty straight forward to implement. For actions that require dropdowns / popovers I will create seperate.

I am currently in the process of tokenising this.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -61,6 +61,14 @@
...
     "./web-apps/*"
   ],
+  "dependencies": {
+    "@tiptap/core": "^2.6.6",
+    "@tiptap/extension-bold": "^2.7.0",
+    "@tiptap/extension-link": "^2.7.2",
+    "@tiptap/extension-underline": "^2.7.1",
+    "@tiptap/pm": "^2.6.6",
+    "@tiptap/starter-kit": "^2.6.6"
+  },
```
Comment (gavin-everett-genesys) [inline package.json:71]: I will remove these they are no longer needed

File: packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor-action-group/gux-rich-text-editor-action-group.scss
Comment (gavin-everett-genesys) [inline packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor-action-group/gux-rich-text-editor-action-group.scss:1]: Will be tokenising this and the other files shortly.

**Comment by gavin-everett-genesys** [comment]
I am going to update this PR to include all the actions that just require a single click since its pretty straight forward to implement. For actions that require dropdowns / popovers I will create seperate.

File: packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/example.html [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+})();
+
+  })()"
+>
+  .prosemirror pre {
+    white-space: pre-wrap;
+  }
+
+  .tiptap {
+    font-size: 14px;
```
Comment (daragh-king-genesys) [inline packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/example.html:120]: What is this styling is every app going to need something like this?

File: packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor-action/gux-rich-text-editor-action.tsx [hunk 3]
```diff
@@ -0,0 +1,67 @@
+import { Component, Element, Prop, h } from '@stencil/core';
+import { trackComponent } from '@utils/tracking/usage';
+
+import { buildI18nForComponent, GetI18nValue } from 'i18n';
+import translationResources from './i18n/en.json';
+import { GuxRichTextEditorActionTypes } from './gux-rich-text-editor-action.types';
+import { calculateDisabledState } from '../gux-rich-text-editor.service';
+import { returnActionTypeIcon } from '../gux-rich-text-editor.service';
```
Comment (daragh-king-genesys) [inline packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor-action/gux-rich-text-editor-action.tsx:8]: These can be joined

File: packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor-action/gux-rich-text-editor-action.tsx [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    }
+  }
+
+  private renderActionButton(): JSX.Element {
+    return (
+      <gux-button-slot accent="ghost" icon-only>
+        <button
+          type="button"
+          disabled={this.disabled}
+          class={{ 'gux-is-active': this.isActive }}
```
Comment (conor-darcy) [inline packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor-action/gux-rich-text-editor-action.tsx:52]: You will probably need aria-selected or aria-pressed here to denote the active state.

**Comment by conor-darcy** [comment]
The icon doesn't seem vertically centred.
image

**Comment by conor-darcy** [comment]
Ordered lists, unordered lists, Block Quotes throwing errors:
Can not convert <> to a Fragment (looks like multiple versions of prosemirror-model were loaded)

**Comment by conor-darcy** [comment]
Should delete clear the content?

**Comment by conor-darcy** [comment]
> Should delete clear the content?

Ignore that, seems to be working & I can't reproduce the issue.

**Comment by gavin-everett-genesys** [comment]
> Can not convert <> to a Fragment (looks like multiple versions of prosemirror-model were loaded)

Looking for a workaround atm, theres a few open issues on github surrounding it 
(https://github.com/ueberdosis/tiptap/issues/3800)

File: packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/example.html [hunk 5]
```diff
@@ -0,0 +1,149 @@
+<h1>gux-rich-text-editor</h1>
+<gux-rich-text-editor-beta>
+  <gux-rich-text-editor-action-group slot="typographical-emphasis">
+    <gux-rich-text-editor-action action="bold"></gux-rich-text-editor-action>
```
[the remark this answers is not in the document]
Comment (gavin-everett-genesys) [reply packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/example.html:4]: In the designs they are referenced as `actions`. Also wanted to keep it consistent with the name we have for actions in the `table-toolbar` as its similar.

File: packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/example.html (hunk 2 above)
Comment (gavin-everett-genesys) [reply packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/example.html:120]: This is tiptap styling but you can also have no styling at all and just have the default editor styles. I have configured it to have the same stylings as the figma designs. Its optional really but might be handy to have it as a template I guess. I could add a comment over it maybe.

**Comment by gavin-everett-genesys** [comment]
> Ordered lists, unordered lists, Block Quotes throwing errors: Can not convert <> to a Fragment (looks like multiple versions of prosemirror-model were loaded)

There seems to a broken CDN import from https://esm.sh. I changed the imports to use jsdelivr instead and it works fine.

(https://github.com/ueberdosis/tiptap/issues/4873)

File: packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/example.html (hunk 2 above)
Comment (daragh-king-genesys) [reply packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/example.html:120]: Should we be including it in a light dom css file that is shipped with the components?

Comment (gavin-everett-genesys) [reply packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/example.html:120]: Yep this makes sense to me 👍

File: packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor-action/gux-rich-text-editor-action.tsx [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+          aria-pressed={this.isActive.toString()}
```
Comment (conor-darcy) [inline packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor-action/gux-rich-text-editor-action.tsx:55]: Do we need an aria-label or some way to describe the button as the icon is decorative?
I know the label is in the tooltip but to got to tab past the button into the tooltip to get the label.

File: packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor.light.scss [hunk 7]
```diff
@@ -0,0 +1,34 @@
+.prosemirror pre {
+  white-space: pre-wrap;
+}
+
```
Comment (conor-darcy) [inline packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor.light.scss:4]: Is there a better class to hang the styles off?
If you use another library you won't get `.prosemirror `or `.tiptap`.

File: packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor-beta.tsx [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    return (
+      <div
+        class={{
+          'gux-rich-text-editor-container': true,
+          'gux-disabled': this.disabled
+        }}
+      >
+        <div class="gux-rich-text-editor-toolbar-container">
+          <div class="gux-typographical-emphasis-container">
+            <slot name="typographical-emphasis"></slot>
```
[the remark this answers is not in the document]
Comment (gavin-everett-genesys) [reply packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor-beta.tsx]: Yeah that makes sense to me 👍

File: packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor-beta.tsx [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+          </div>
+          <div class="gux-text-styling-container">
+            <slot name="text-styling"></slot>
```
[the remark this answers is not in the document]
Comment (gavin-everett-genesys) [reply packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor-beta.tsx]: In the designs in figma it states that each group should have its own slot. I am open to change but thinking how the responsive behaviour will work also. At the moment I can use the wrapper classes potentially for hiding and showing those action groups whereas if were just using a single slot I am not sure how that will work. Open to suggestions

File: packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor-action/gux-rich-text-editor-action.tsx (hunk 6 above)
Comment (gavin-everett-genesys) [reply packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor-action/gux-rich-text-editor-action.tsx:55]: Think the use of `screenreader-text` property on the the icon will suffice 🤔

File: packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor.light.scss (hunk 7 above)
Comment (gavin-everett-genesys) [reply packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor.light.scss]: Good spot, I am now just applying an `editor-styles` class to the editor on load. I assume other editors might have the same functionality where they can add a class.

File: packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor-action-group/gux-rich-text-editor-action-group.tsx [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+  private isGlobalActionsSlot(): boolean {
+    return this.root.getAttribute('slot') === 'global-actions';
+  }
+
+  private renderActionGroupDivider(): JSX.Element {
+    if (!this.shouldHideDivider()) {
+      return (
+        <div class="gux-action-group-divider">
+          <div class="gux-divider"></div>
+        </div>
```
[the remark this answers is not in the document]
Comment (gavin-everett-genesys) [reply packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor-action-group/gux-rich-text-editor-action-group.tsx:32]: Yeah thats true I could just apply a padding to the left and right of the `gux-divider` and just have one div but I was matching it to the design in figma where it shows it like this,
Screenshot 2024-10-21 at 08 59 58

but then in the example of the toolbar it shows it with a padding left of 8px and padding-right of 8px so I will just shorten it to one div.

[the remark this answers is not in the document]
Comment (conor-darcy) [reply packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor-action-group/gux-rich-text-editor-action-group.tsx]: @gavin-everett-genesys Fair enough, just work with the tokens you have.

File: packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor-action/gux-rich-text-editor-action.tsx (hunk 6 above)
Comment (conor-darcy) [reply packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor-action/gux-rich-text-editor-action.tsx:55]: I checked it with Voiceover & it announces the label ok.

File: packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor-action-group/gux-rich-text-editor-action-group.scss [hunk 11]
```diff
@@ -0,0 +1,26 @@
+:host {
+  display: flex;
+  flex-direction: row;
+}
+
+:host([slot='global-actions']) {
+  ::slotted(gux-rich-text-editor-action:last-child) {
+    margin-left: 20px; //Need to confirm with UX.
```
Comment (conor-darcy) [inline packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor-action-group/gux-rich-text-editor-action-group.scss:8]: Don't forget this or create a separate ticket.

File: packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor.scss [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    justify-content: flex-start;
+  }
+
+  &.gux-disabled {
+    pointer-events: none;
+    cursor: default;
+    user-select: none;
+    opacity: 0.5;
+
+    .tiptap {
```
Comment (conor-darcy) [inline packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor.scss:32]: I don't think we should reference tiptap here.

**Review by conor-darcy — COMMENTED** [review]
Whats your plan with tokens?

File: packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor-action/gux-rich-text-editor-action.tsx [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+  async componentWillLoad(): Promise<void> {
+    trackComponent(this.root, { variant: this.action });
+    this.i18n = await buildI18nForComponent(this.root, translationResources);
+    this.disabled = calculateDisabledState(this.root);
+  }
+
+  private renderTooltip(): JSX.Element {
+    if (!this.disabled) {
+      return (
+        <gux-tooltip>
```
Comment (321gillian) [inline packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor-action/gux-rich-text-editor-action.tsx:41]: any reason to prefer gux-tooltip over the beta version here?

File: packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor-action/i18n/en.json [hunk 14]
```diff
@@ -0,0 +1,14 @@
+{
+  "bold": "Bold",
+  "italic": "Italic",
+  "underline": "Underline",
+  "strike": "Strike",
+  "codeblock": "Code Block",
+  "blockQuote": "Block Quote",
+  "orderedList": "Ordered List",
+  "bulletList": "Bullet List",
```
Comment (321gillian) [inline packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor-action/i18n/en.json:9]:
```suggestion
  "bulletList": "Bulleted List",
```

File: packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor-action/gux-rich-text-editor-action.tsx (hunk 13 above)
Comment (gavin-everett-genesys) [reply packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor-action/gux-rich-text-editor-action.tsx:41]: No preference really will use the beta version.

File: packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor.light.scss [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+  font-family: var(--gse-ui-rte-paragraph-md-regular-fontFamily);
+  font-size: var(--gse-ui-rte-paragraph-md-regular-fontSize);
+  font-weight: var(--gse-ui-rte-paragraph-md-regular-fontWeight);
+  line-height: var(--gse-ui-rte-paragraph-md-regular-lineHeight);
+  color: var(--gse-ui-rte-paragraph-foregroundColor);
+  outline: none;
+
+  pre {
+    white-space: pre-wrap;
+  }
```
Comment (321gillian) [inline packages/genesys-spark-components/src/components/beta/gux-rich-text-editor/gux-rich-text-editor.light.scss:11]: nitpick, there's two pre style blocks here

File: packages/genesys-spark-tokens/reports/used-undefined-tokens.json [hunk 16]
```diff
@@ -1 +1 @@
-[]
+["--gse-ui-button-icon-only-width", "--gse-ui-rte-codeblock-border-color"]
```
Comment (321gillian) [inline packages/genesys-spark-tokens/reports/used-undefined-tokens.json:1]: does this need to be resolved?

Comment (gavin-everett-genesys) [reply packages/genesys-spark-tokens/reports/used-undefined-tokens.json]: Think this occured due to the token changes so `--gse-ui-button-icon-only-width` became `--gse-ui-button-iconOnly-width` and  `--gse-ui-rte-codeblock-border-color` the `b` needed to be capitalised `--gse-ui-rte-codeBlock-border-color`

**Review by 321gillian — APPROVED** [review]

**Pull request merged**